### PR TITLE
Add support for deferred sampling

### PR DESCRIPTION
**Pull request by vprithvi**
- When a downstream service doesn't have enough information to
  make a sampling decision, it sets the span context flag 4.

- When a span with the flag 4 is encountered, a sampling decision
  for any child spans is made based on the current operation name
  instead of depending on the sampling decision on the parent if
  the span context doesn't contain the debug flag(2). Debug spans
  are always sampled.

- set the tag deferredsampling.status to indicate that the sampling
  decision was taken at this span

### CODE REVIEW

File: src/tracer.js [hunk 1]
```diff
@@ -231,13 +231,24 @@ export default class Tracer {
...
             ctx.baggage = parent.baggage;
 
-            parent.finalizeSampling();
-            ctx.finalizeSampling();
+            // Parent exists without a sampling decision
+            // An example of where this happens is in mobile clients where sampler parameters
+            // are not propagated
+            if ((parent.flags & constants.DEFERRED_SAMPLING_MASK)
+                === constants.DEFERRED_SAMPLING_MASK) {
+                ctx.flags = parent.flags ^ constants.DEFERRED_SAMPLING_MASK
```
Comment (black-adder) [inline src/tracer.js]: I think we need to remove the deferred_sampling_mask in the parent.flag too and also set the sampled_mask on it too if it should be sampled. This is because the same parent context can be used to create multiple spans. We don't want to run this logic multiple times, only once on the first span using this parent span. Future spans will just respect the parent context which has been modified correctly.

File: src/tracer.js [hunk 2]
```diff
@@ -231,13 +231,24 @@ export default class Tracer {
...
+                if (this._sampler.isSampled(operationName, internalTags)) {
+                    ctx.flags = parent.flags |= constants.SAMPLED_MASK
```
Comment (black-adder) [inline src/tracer.js]: I prefer if we don't use parent.flags here because all its pertinent information has already been copied into ctx.flags.

File: src/tracer.js [hunk 3]
```diff
@@ -231,13 +231,24 @@ export default class Tracer {
...
+                    internalTags[constants.DEFERRED_SAMPLING_STATUS] = true
+                }
+            } else {
+                ctx.flags = parent.flags
+                parent.finalizeSampling();
```
Comment (black-adder) [inline src/tracer.js]: This also needs to be called in the if case above.

File: test/span.js [hunk 4]
```diff
@@ -336,3 +336,64 @@ describe('span should', () => {
     // TODO(oibe) need tests for standard tags, and handlers
 });
 
+describe('span with deferred sampling flag', () => {
```
Comment (black-adder) [inline test/span.js]: add a test where two spans are created from one deferred parent span.

File: src/tracer.js [hunk 5]
```diff
@@ -231,13 +231,24 @@ export default class Tracer {
...
-
             // reuse parent's baggage as we'll never change it
             ctx.baggage = parent.baggage;
 
-            parent.finalizeSampling();
-            ctx.finalizeSampling();
+            // Parent exists without a sampling decision
+            // An example of where this happens is in mobile clients where sampler parameters
+            // are not propagated
+            if ((parent.flags & constants.DEFERRED_SAMPLING_MASK)
```
Comment (black-adder) [inline src/tracer.js]: nit: can we simply do
`if (parent.flags & constants.DEFERRED_SAMPLING_MASK)`
because all we need to do is confirm flag & mask is not 0.

File: src/constants.js [hunk 6]
```diff
@@ -24,12 +24,18 @@ export const SAMPLED_MASK = 0x1;
...
+export const DEFERRED_SAMPLING_MASK = 0x4;
+
 // JAEGER_CLIENT_VERSION_TAG_KEY is the name of the tag used to report client version.
 export const JAEGER_CLIENT_VERSION_TAG_KEY = 'jaeger.version';
 
 // TRACER_HOSTNAME_TAG_KEY used to report host name of the process.
 export const TRACER_HOSTNAME_TAG_KEY = 'jaeger.hostname';
 
+// DEFERRED_SAMPLING_STATUS indicated that deferred sampling was performed at this span.
+export const DEFERRED_SAMPLING_STATUS = 'deferredsampling.status'
```
Comment (vprithvi) [inline src/constants.js]: How about `sampler.deferred`, which makes it consistent with the other sampler tags. Essentially this indicates that sampling has been deferred until now

File: src/tracer.js (hunk 5 above)
Comment (vprithvi) [inline src/tracer.js]: Done

File: src/constants.js [hunk 7]
```diff
@@ -24,12 +24,18 @@ export const SAMPLED_MASK = 0x1;
 // DEBUG_MASK is the bit mask indicationg that a span has been marked for debug.
 export const DEBUG_MASK = 0x2;
 
+// DEFERRED_SAMPLING_MASK is the bit mask indicating that a span has been marked for deferred sampling.
```
Comment (yurishkuro) [inline src/constants.js]: better to elaborate what deferred sampling means

File: src/constants.js (hunk 6 above)
Comment (yurishkuro) [inline src/constants.js]: first, if we want to use this as a tag, then the const naming pattern is XXX_TAG_KEY (see  JAEGER_CLIENT_VERSION_TAG_KEY, and its description).

But I don't see much need for this tag in the first place. What is the intended use case? The "deferred sampling was performed at this span" sounds somewhat strange to me, because "deferring" the sampling decision actually happened on the mobile side.

Comment (vprithvi) [inline src/constants.js]: I was thinking that adaptive sampling might have to account for deferred sampling traffic separately because the sampling decision is taking place on a non-root span. 
Essentially, I wanted to indicate that the deferred sampling decision was converted to a concrete sampling decision at this span.

File: src/tracer.js (hunk 2 above)
Comment (vprithvi) [inline src/tracer.js]: Done

File: src/constants.js (hunk 6 above)
Comment (black-adder) [inline src/constants.js]: I think the plan is to use the sampler.type / sampler.param flags instead to determine whether this span was sampled rather than using root span. This way we don't need this extra tag (I'm basing this off of the discussion we had in the mobile RFC).

However, I do like having this extra tag even if it's not useful for adaptive sampling. For audit purposes.

File: test/span.js (hunk 4 above)
Comment (vprithvi) [inline test/span.js]: Done

File: src/constants.js (hunk 7 above)
Comment (vprithvi) [inline src/constants.js]: Done

File: src/tracer.js (hunk 1 above)
Comment (vprithvi) [inline src/tracer.js]: Agree about running the same thing multiple times, but as Yuri said, I think that we shouldn't be modifying the parent span because it was created elsewhere.

File: src/tracer.js (hunk 3 above)
Comment (vprithvi) [inline src/tracer.js]: Why is that? How do we know that `setOperation` isn't going to be called on the parent?
I think makes sense in the else, because a sampling decision is already made.

File: src/tracer.js [hunk 8]
```diff
@@ -231,13 +231,23 @@ export default class Tracer {
...
             // reuse parent's baggage as we'll never change it
             ctx.baggage = parent.baggage;
+            ctx.flags = parent.flags;
 
-            parent.finalizeSampling();
-            ctx.finalizeSampling();
+            // Parent exists without a sampling decision
+            // An example of where this happens is in mobile clients where sampler parameters
+            // are not propagated
+            if (parent.flags & constants.DEFERRED_SAMPLING_MASK) {
```
Comment (yurishkuro) [inline src/tracer.js]: ctx.flags ?

File: src/tracer.js [hunk 9]
```diff
@@ -231,13 +231,23 @@ export default class Tracer {
...
+                ctx.flags ^=  constants.DEFERRED_SAMPLING_MASK;
+                if (this._sampler.isSampled(operationName, internalTags)) {
+                    ctx.flags |= constants.SAMPLED_MASK;
+                    internalTags[constants.DEFERRED_SAMPLING_TAG_KEY] = true;
+                }
+            } else {
+                parent.finalizeSampling();
+                ctx.finalizeSampling();
```
Comment (yurishkuro) [inline src/tracer.js]: I would like to consider whether "finalized" and "deferred" are actually the same thing.

File: src/tracer.js (hunk 8 above)
Comment (black-adder) [inline src/tracer.js]: This still doesn't solve what happens if the parent context creates more than 1 child span. The tracer might be using a ratelimiting sampler which means that sampler.isSampled might return different results for 2 spans in the same trace which is not what we want.

Comment (vprithvi) [inline src/tracer.js]: Done

File: src/tracer.js (hunk 9 above)
Comment (vprithvi) [inline src/tracer.js]: I don't think they are the same thing. 

I was also looking at @black-adder's comment about this scheme not working with the rate limiting sampler (or in the more general case, any sampler that isn't consistent). The previous suggestion of modifying the parent span solves this problem. 

The `finalizeSampling` mechanism resets the sampling decision by calling `IsSampled` the first time `setOperation` is called.  Any subsequent `setOperation` calls are ignored.

Further, the `finalizeSampling` mechanism is only invoked for new root spans. (Anything extracted/injected is finalized in the extractor/injector)

I feel that @black-adder's suggestion might be the cleanest way of sampling non root spans in the middle of a trace.

Comment (yurishkuro) [inline src/tracer.js]: You are referring to the _current implementation_ of finalized sampling. I want to step back and consider what the actual difference is between deferred and not finalized - even from their name they appear to be very similar. The only difference I currently see is deferred works across services, while not-finalized within a single span. But I do think we can generalize this to a single concept and simplify the code.

Ultimately they both defer the sampling decision until some event later in the execution.

File: src/constants.js (hunk 7 above)
Comment (yurishkuro) [inline src/constants.js]: the first sentence is a bit of a tautology. 

DEFERRED_SAMPLING_MASK is the bit mask indicating that the upstream service that generated the span did not make a definitive sampling decision, but deferred it to be made by some downstream service.

No need to mention mobile, imo, especially since the claim that we cannot have a feedback loop with them is not exactly true.

File: src/tracer.js [hunk 10]
```diff
@@ -228,6 +228,15 @@ export default class Tracer {
             ctx.parentId = null;
             ctx.flags = flags;
         } else {
+            if (!parent.samplingFinalized) {
+                if (this._sampler.isSampled(operationName, internalTags)) {
+                   parent._flags |= constants.SAMPLED_MASK
+                } else {
+                    parent._flags &= ~constants.SAMPLED_MASK
+                }
+                parent.finalizeSampling()
```
Comment (yurishkuro) [inline src/tracer.js]: don't like this. Parent context should be treated as immutable. Often it will be deserialized from the wire, so trying to modify it is misleading at best.

File: src/tracer.js [hunk 11]
```diff
@@ -209,7 +209,7 @@ export default class Tracer {
         let internalTags: any = {};
         if (!parent || !parent.isValid) {
```
Comment (yurishkuro) [inline src/tracer.js]: I think this could be refactored to avoid repetition. For example, the invalid parent behavior can be done upfront:

```
if (parent && !parent.isValid()) {
   // ...
   parent = null;
}
```

then create new Context and update it depending on parent == null

then check if context is deferred and make sampling decision

File: src/tracer.js [hunk 12]
```diff
@@ -297,12 +305,7 @@ export default class Tracer {
             throw new Error(`Unsupported format: ${format}`);
         }
 
-        let spanContext = extractor.extract(carrier);
-
-        if (spanContext) {
-            spanContext.finalizeSampling();
-        }
-        return spanContext;
+        return extractor.extract(carrier);
```
Comment (yurishkuro) [inline src/tracer.js]: hm, not sure why it was doing that before. Has any test broke because of this?

Comment (vprithvi) [inline src/tracer.js]: No, none of the tests break because of this. In this implementation,  finalizing a span removes the deferred sampling flag, so finalizing a span before we apply a sampling decision is problematic.

File: src/tracer.js (hunk 10 above)
Comment (vprithvi) [inline src/tracer.js]: We could have another boolean in the parent span to store the state of whether all child spans must be sampled and update that instead of the parent span context. Would you prefer the implementation to do that?

File: src/constants.js (hunk 7 above)
Comment (vprithvi) [inline src/constants.js]: I'll update the first sentence and remove the example about mobile. 
However, I haven't claimed that such a feedback loop cannot exist.

Comment (yurishkuro) [inline src/constants.js]: > because adaptive sampling parameters cannot be reliably pushed to mobile devices

File: src/tracer.js (hunk 10 above)
Comment (yurishkuro) [inline src/tracer.js]: we do not have access to parent span when creating a child span, we only have access to its context

File: src/constants.js (hunk 7 above)
Comment (vprithvi) [inline src/constants.js]: Well, that statement is akin to saying that the network isn't reliable, and thus the feedback loop might not be reliable as well. It does not claim that we cannot have a feedback loop.

File: src/tracer.js (hunk 11 above)
Comment (vprithvi) [inline src/tracer.js]: It becomes a bit messy with handling the debug spans as well. I'd prefer this refactor be a part of a separate commit.

File: src/tracer.js (hunk 10 above)
Comment (vprithvi) [inline src/tracer.js]: Indeed, in which case modifying the parent seems reasonable.

File: src/span_context.js [hunk 13]
```diff
@@ -126,8 +108,25 @@ export default class SpanContext {
...
+     * - it has inherited the sampling decision from its parent
+     * - its debug flag is set via the sampling.priority tag
+     * - it is finish()-ed
+     * - setOperationName is called
+     * - it is used as a parent for another span
+     * - its context is serialized using injectors
+     * */
     get samplingFinalized(): boolean {
-        return this._samplingFinalized;
+        return ((this.flags & constants.DEFERRED_SAMPLING_MASK) !== constants.DEFERRED_SAMPLING_MASK)
```
Comment (yurishkuro) [inline src/span_context.js]: - outer parentheses not needed
- missing `;`

File: src/span_context.js [hunk 14]
```diff
@@ -162,7 +161,7 @@ export default class SpanContext {
     }
 
     finalizeSampling(): void {
-        this._samplingFinalized = true;
+        this.flags &= ~constants.DEFERRED_SAMPLING_MASK
```
Comment (yurishkuro) [inline src/span_context.js]: add `;`

File: src/span.js [hunk 15]
```diff
@@ -118,12 +117,21 @@ export default class Span {
...
     context(): SpanContext {
-        return this._spanContext;
+        let ctx = this._spanContext;
+        if (!ctx.samplingFinalized) {
+            if (ctx.isDebug() || this._tracer._sampler.isSampled(this.operationName, this._tags)) {
+                ctx._flags |= constants.SAMPLED_MASK
+            } else {
+                ctx._flags &= ~constants.SAMPLED_MASK
+            }
+            ctx.finalizeSampling()
```
Comment (yurishkuro) [inline src/span.js]: add `;` everywhere

File: src/tracer.js [hunk 16]
```diff
@@ -228,6 +228,7 @@ export default class Tracer {
             ctx.parentId = null;
             ctx.flags = flags;
         } else {
+            this.processDeferredSampling(parent, operationName, internalTags);
```
Comment (yurishkuro) [inline src/tracer.js]: I think this should be called after L235 and operate on `ctx`, not on `parent`.

Comment (vprithvi) [inline src/tracer.js]: The problem with the approach you suggest is that we can no longer guarantee the consistency of the sampling decision for all children (because each child has an independent sampling decision). We need to store the sampling decision for all future spans in the parent span context object or something associated with it to ensure all child spans have the same sampling decision applied to them.

I don't think that this is a easy problem to solve neatly while most of our stack assumes that sampling decisions happen when the span context is initially created.

Comment (yurishkuro) [inline src/tracer.js]: Since we narrowed the scope, would't the `parent` in this case always be something we just serialized from the wire? The _only_ child is going to be the server-side span.

And as a more general principle, if the parent wants to ensure that its sampling decision is fixed and should affect all children, then it's he parent who should make the change, not as a side effect of creating a child.

Comment (vprithvi) [inline src/tracer.js]: >Since we narrowed the scope, would't the parent in this case always be something we just serialized from the wire? The only child is going to be the server-side span.
You are right, I'll have an updated diff out shortly
